File: csharp/solution_and_tests/two_pointers/problem_0016_3sum_closest.cs
/*
 * https://leetcode.com/problems/3sum-closest/description/ 
 * Explanation:
 * 
 * Similar to 3Sum, we would use the two pointers technique here. This requires a sorted Array
 * However, since this ideal triplet may not exist, we will track the smallest absolute difference
 * between the threeSum and the target
 * 
 * If diff is 0, this is our ideal triplet, break from the loop
 *
 * Time: O(n log n + n ^2) = O(n^2)
 * Space: O(log n) to O(n), depending on the implementation of the sorting algorithm
 */

using System;
using NUnit.Framework;

namespace ThreeSum_Closest
{
    class Solution
    {
        public int ThreeSumClosest(int[] nums, int target)
        {
            Array.Sort(nums);
            int diff = Int32.MaxValue;
            for (int i = 0; i < nums.Length && diff != 0; i++)
            {
                int leftPointer = i + 1;
                int rightPointer = nums.Length - 1;
                while (leftPointer < rightPointer)
                {
                    int threeSum = nums[i] + nums[leftPointer] + nums[rightPointer];
                    int closestThreeSum = target - threeSum;
                    if (Math.Abs(closestThreeSum) < Math.Abs(diff))
                    {
                        diff = closestThreeSum;
                    }
                    if (threeSum < target)
                    {
                        leftPointer++;
                    }
                    else
                    {
                        rightPointer--;
                    }
                }
            }
            return target - diff;
        }
    }

    class Tests
    {
        Solution sol;
        [SetUp]
        public void SetUp()
        {
            sol = new Solution();
        }

        [Test]
        [TestCase(new int[] { -1, 2, 1, -4 }, 1, ExpectedResult = 2)]
        [TestCase(new int[] { 0, 0, 0 }, 1, ExpectedResult = 0)]
        [TestCase(new int[] { 4, 0, 5, -5, 3, 3, 0, -4, -5 }, -2, ExpectedResult = -2)]
        public int TestThreeSumClosest(int[] nums, int target)
        {
            return sol.ThreeSumClosest(nums, target);
        }
    }
}

File: JavaMaven/src/test/java/two_pointers/problem_0016_3sum_closest.java
/*
 * https://leetcode.com/problems/3sum-closest/description/ 
 * Explanation:
 * 
 * Similar to 3Sum, we would use the two pointers technique here. This requires a sorted Array
 * However, since this ideal triplet may not exist, we will track the smallest absolute difference
 * between the threeSum and the target
 * 
 * If diff is 0, this is our ideal triplet, break from the loop
 *
 * Time: O(n log n + n ^2) = O(n^2)
 * Space: O(log n) to O(n), depending on the implementation of the sorting algorithm
 */

package two_pointers;

import java.util.Arrays;

import org.junit.Test;
import static org.junit.Assert.assertEquals;

public class problem_0016_3sum_closest {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int diff = Integer.MAX_VALUE;
        for (int i = 0; i < nums.length && diff != 0; i++) {
            int left = i + 1;
            int right = nums.length - 1;
            while (left < right) {
                int threeSum = nums[i] + nums[left] + nums[right];
                int closestThreeSum = target - threeSum;
                if (Math.abs(diff) > Math.abs(closestThreeSum)) {
                    diff = closestThreeSum;
                }
                if (threeSum < target) {
                    left++;
                } else {
                    right--;
                }

            }
        }
        return target - diff;
    }

    @Test
    public void testThreeSumClosest() {
        int[] testCase1 = new int[] { -1, 2, 1 - 4 };
        int testCase1Target = 1;
        int testCase1ExpectedResult = -2;
        int[] testCase2 = new int[] { -1, 2, 1 - 4 };
        int testCase2Target = 0;
        int testCase2ExpectedResult = -2;
        int[] testCase3 = new int[] { 4, 0, 5, -5, 3, 3, 0, -4, -5 };
        int testCase3Target = -2;
        int testCase3ExpectedResult = -2;

        assertEquals(testCase1ExpectedResult, threeSumClosest(testCase1, testCase1Target));
        assertEquals(testCase2ExpectedResult, threeSumClosest(testCase2, testCase2Target));
        assertEquals(testCase3ExpectedResult, threeSumClosest(testCase3, testCase3Target));
    }
}